### ISSUE TITLE
Is NullPointerException here the right thing?

### ISSUE THREAD

**Issue by StefanOltmann** [pitschr/knx-link#104]
Hi Christoph.

I just tested the behaviour of knx-link in an environment without KNX.

In that case I receive a `NullPointerException`. 

I once learned that a NPE should only be thrown by the JVM itself and otherwise be avoided.

So for checking arguments for example I always throw an `IllegalArgumentException` if it's null (or has another illegal value).

In this case I think there might be room for a specialized exception that tells me that nothing with the framework is wrong (NullPointerExceptions look to me like something internal broke), but there is just nothing to connect to.

From the outside I could catch that specific exception and react to this situation more specific than I would do it if knx-link encounters really unexpected problem.

I see that there is a `KnxDiscoveryNotReceivedException` in the stack trace also. I guess for a better experience that should be the very first thing the stack trace starts with.

What do you think?

Full stack trace:
```
java.lang.NullPointerException: No search response received for request: SearchRequestBody{discoveryEndpoint=HPAI{length=8 (0x08), protocol=HostProtocol{name=IPV4_UDP, friendlyName=IP v4 UDP communication, code=1 (0x01)}, address=0.0.0.0 (0x00 00 00 00), port=3671 (0x0E 57)}, rawData=0x08 01 00 00 00 00 0E 57}
	at li.pitschmann.knx.core.utils.Preconditions.checkNonNull(Preconditions.java:63)
	at li.pitschmann.knx.core.communication.InternalKnxClient.fetchDiscoveryFromKNX(InternalKnxClient.java:535)
	at li.pitschmann.knx.core.communication.InternalKnxClient.startTunneling(InternalKnxClient.java:165)
	at li.pitschmann.knx.core.communication.InternalKnxClient.start(InternalKnxClient.java:125)
	at li.pitschmann.knx.core.communication.DefaultKnxClient.createStarted(DefaultKnxClient.java:78)
	at de.stefan_oltmann.smarthome.server.knx.KnxClient.<init>(KnxClient.kt:67)
```

**Comment by pitschr** [comment]
Hi Stefan.

NPE has always been a debatable topic 😄 . NPE is fine when you expect that the object is non-null but was null at this point. NPE should not be exclusively for JVM only. Is this argument coming from Kotlin community 😓 ? Almost all java libraries including Java API are doing that way.

`IllegalArgumentException` is usually thrown when you provided an illegal non-null argument, this is not the case here. Better would be probably using the `IllegalStateException`? But a customized exception is usually the cherry on the top.

Come back to the topic. The issue you face is the exception in the logging, right? Your client should get a `KnxDiscoveryNotReceivedException` rather than `NPE` itself. I think here it is sufficient to omit the logging line as it is somehow duplicate and part of the `KnxDiscoveryNotReceivedException` anyway. I think that should be a good way and I'll apply it same for `InternalKnxClient#fetchChannelIdFromKNX()`. Thanks.

**Comment by StefanOltmann** [comment]
> NPE has always been a debatable topic 😄 . NPE is fine when you expect that the object is non-null but was null at this point. NPE should not be exclusively for JVM only. Is this argument coming from Kotlin community 😓 ? Almost all java libraries including Java API are doing that way.

If I remember right I read it back then in the "SCJP Sun Certified Programmer for Java 6 Study Guide".

> `IllegalArgumentException` is usually thrown when you provided an illegal non-null argument, this is not the case here. 

I personally throw that for any kind of illegal argument, mostly nulls. But that may be personal taste.

> Better would be probably using the `IllegalStateException`? But a customized exception is usually the cherry on the top.

`IllegalStateException` in my view is when an attribute of an object has a wrong value. I use that exception if something has not been initialized because a prior requiered call to a method did not occur and I figure that out elsewhere (where I need that). This way I trace if my dependency injection is messed up. But that also may be a debatable topic. :)

> Come back to the topic. The issue you face is the exception in the logging, right?

Yes. What I quoted above is the beginning of the stack trace. The `KnxDiscoveryNotReceivedException` comes only later if I scroll down.

> Your client should get a `KnxDiscoveryNotReceivedException` rather than `NPE` itself. I think here it is sufficient to omit the logging line as it is somehow duplicate and part of the `KnxDiscoveryNotReceivedException` anyway. I think that should be a good way and I'll apply it same for `InternalKnxClient#fetchChannelIdFromKNX()`. Thanks.

Yes, you may try that. I did not test if that's the solution. My suggestion is to build it the way that a meaningful exception is the first thing I see.

I don't know if you can detect that, but a `NoKnxDevicesPresentException` would be the most clear exception. But I guess you can only tell that by getting no response to the discovery request - so that  `KnxDiscoveryNotReceivedException` is accurate.

**Comment by StefanOltmann** [comment]
Oh, there really is a debate :D

https://www.baeldung.com/java-illegalargumentexception-or-nullpointerexception

After over a decade programming Java one of my reasons to switch to Kotlin is the nullsafety. If done right you very rarely face an NPE anymore. 👍

**Comment by pitschr** [comment]
Yeah. I personally rather stick with the official way whenever possible.

Btw. the issue is now fixed and is a part of knx-core version 0.3.0 which is has been released.
https://search.maven.org/artifact/li.pitschmann/knx-core/0.3.0/jar

FYI I now splitted the project "knx-link" into multiple projects "knx-core", "knx-core-plugins" and "knx-logic" (and logic sub-modules as well, WIP). The reason is that the project "knx-link" became too big meanwhile with its modules and it is much bigger than I planned initially. From maven perspective it was also not good when I released new artifacts although they have not been touched at all. Most important: **From maven perspective it should be no difference, your gradle/pom remain unchanged.** The changes are to my GIT repositories and I'm also more flexible including independent release process.

https://github.com/pitschr/knx-core   (incl. updated readme in some parts)
https://github.com/pitschr/knx-core-plugins
https://github.com/pitschr/knx-logic

So "knx-link" wont be in your interest anymore, I assume. You may unwatch "knx-link" and watch "knx-core" instead. Please raise your issues in "knx-core" in future. Thanks!

Not sure what happens to "knx-link" repo. For the "knx-link" I have the vision to create a systemd service which is Java-based with "knx-core" as dependency. For CLI tool I am probably choosing Rust as native to outperform the cold starts that comes with JVM. This CLI will then communicate with the daemon process. It is also a great opportunity for me to learn the Rust language.

**Comment by StefanOltmann** [comment]
Good decision to split the repo. 👍 

I like your plan on going native. I also try to build my REST API native with Quarkus and GraalVM native-image, but I struggle to get around that `Inet4Address` problem for now.

If you manage to make this a native CLI tool that crappy knxd can finally die. ❤️
